### PR TITLE
Story 2614. Incremented major version and added new-commands-README.m…

### PR DESCRIPTION
**Pull request by yeunga**
…d file.

### CODE REVIEW

File: vos/vos/commands/vcp.py [hunk 1]
```diff
@@ -355,25 +382,40 @@ def copy(source_name, destination_name, exclude=None, include=None,
...
+                logging.debug('Source {} can be accessed.'.format(source))
+
                 if not args.follow_links and islink(source):
                     logging.info("{}: Skipping (symbolic link)".format(source))
                     continue
 
                 # copying inside VOSpace not yet implemented
-                if source[0:4] == 'vos:' and dest[0:4] == 'vos:':
+                if (_is_vos(source) or vos.is_remote_file(source))\
+                        and (_is_vos(dest) or vos.is_remote_file(dest)):
```
Comment (andamian) [inline vos/vos/commands/vcp.py]: I find this `_is_vos`, `_is_remote_file`, `is_remote` confusing. There should be only one remote in the client. For vos client remote means `vos` in scheme. For inventory storage means scheme not `file:`. 
This check will be to see if both source and destination are remote.

File: vos/vos/commands/vrm.py [hunk 2]
```diff
@@ -1,42 +1,78 @@
...
 """remove a vospace data node, fails if container node or node is locked."""
 import logging
+import six.moves.urllib.parse as urllibparse
 from ..commonparser import set_logging_level_from_args, exit_on_exception, \
     CommonParser
-from .. import vos
+from .. import vos, storage_inventory
 
-DESCRIPTION = """remove a vospace data node; fails if container node or node is locked.
+urlparse = urllibparse.urlparse
```
Comment (andamian) [inline vos/vos/commands/vrm.py]: same comment as for vcp

File: vos/vos/commands/vrm.py [hunk 3]
```diff
@@ -1,42 +1,78 @@
...
+node or node is locked.
 
 eg. vrm vos:/root/node   -- deletes a data node"""
 
 
+def delete_nodes(args):
+    client = vos.Client(vospace_certfile=args.certfile,
+                        vospace_token=args.token)
+    for node in args.source:
+        if not node.startswith('vos:'):
```
Comment (andamian) [inline vos/vos/commands/vrm.py]: Replace with `_is_vos` or `_is_remote`

File: vos/vos/commands/vrm.py [hunk 4]
```diff
@@ -1,42 +1,78 @@
...
+            raise Exception(
+                '{} is not a valid VOSpace handle'.format(node))
+        if not node.endswith('/'):
+            if client.get_node(node).islink():
```
Comment (andamian) [inline vos/vos/commands/vrm.py]: `get_node` is an extra overhead. Is it required? Can we get the node/link info from the response? Do we need it?

File: vos/vos/commands/vrm.py [hunk 5]
```diff
@@ -1,42 +1,78 @@
...
+                logging.info('deleting link {}'.format(node))
+                client.delete(node)
+            elif client.isfile(node):
+                logging.info('deleting {}'.format(node))
+                client.delete(node)
+        elif client.isdir(node):
+            raise Exception('{} is a directory'.format(node))
+        else:
+            raise Exception('{} is not a directory'.format(node))
```
Comment (andamian) [inline vos/vos/commands/vrm.py]: What is this message? What does it mean for the user? Is this a valid case or a BUG?

File: vos/vos/commands/vrm.py [hunk 6]
```diff
@@ -1,42 +1,78 @@
...
+
+
+def delete_files(args):
+    client = storage_inventory.Client(args.resource_id,
+                                      certfile=args.certfile,
+                                      token=args.token)
+    for file in args.source:
+        if file.startswith('vos:'):
+            raise Exception(
+                '{} is not a valid storage handle'.format(file))
```
Comment (andamian) [inline vos/vos/commands/vrm.py]: In theory nothing should stop someone from using a `vos` URI with storage. In practice, this check is probably valid.

File: vos/vos/commands/vrm.py [hunk 7]
```diff
@@ -1,42 +1,78 @@
...
+            elif client.isfile(node):
+                logging.info('deleting {}'.format(node))
+                client.delete(node)
+        elif client.isdir(node):
+            raise Exception('{} is a directory'.format(node))
+        else:
+            raise Exception('{} is not a directory'.format(node))
+
+
+def delete_files(args):
```
Comment (andamian) [inline vos/vos/commands/vrm.py]: If it was me, I would create the appropriate client up front and call one delete (different implementations) on it after determining that the arg is not a directory.

File: vos/vos/commands/vrm.py [hunk 8]
```diff
@@ -1,42 +1,78 @@
...
     parser = CommonParser(description=DESCRIPTION)
-    parser.add_argument('node',
-                        help='dataNode or linkNode to delete from VOSpace',
-                        nargs='+')
+    parser.add_argument(
+        "--resource-id", default=None,
+        help="resource ID of the Storage Inventory service to be used")
+    parser.add_argument(
+        'source',
+        help='file, dataNode or linkNode to delete from VOSpace',
```
Comment (andamian) [inline vos/vos/commands/vrm.py]: artifact/resource to delete.

File: vos/vos/storage_inventory.py [hunk 9]
```diff
@@ -0,0 +1,315 @@
+"""A set of Python Classes for connecting to and interacting with a VOSpace
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: VOSpace?

File: vos/vos/storage_inventory.py [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+
+urlparse = parse.urlparse
+
+# This is an encoder.  It will URL encode any URI component.
+# It is used in creating the DELETE URL by appending the
+# encoded URI to the end of the path.
+#
+quote_plus = parse.quote_plus
+
+logger = logging.getLogger('vos')
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: Replace `vos` with `__name__`

File: vos/vos/storage_inventory.py [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+logger.setLevel(logging.ERROR)
+
+# Header for token access
+HEADER_DELEG_TOKEN = 'X-CADC-DelegationToken'
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: I don't think it's needed

File: vos/vos/storage_inventory.py [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+        :param resource_id:     The service Resource ID to use.  Defaults to
+        the Global Raven Site.
+        :param certfile: x509 proxy certificate file location.
+        Overrides certfile in conn.
+        :type certfile: unicode
+        :param token: token string (alternative to certfile)
+        :type token: unicode
+        :param conn: a connection pool object for this Client
+        :type conn: Session
+        :param http_debug: turn on http debugging.
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: remove

File: vos/vos/storage_inventory.py [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+        """
+            Copy from the source to the destination.
+
+            One of source or destination must be a vospace location and the
+            other must be a local location.
+
+            :param source: The source file to send to VOSpace or the VOSpace
+            node to retrieve
+            :type source: unicode
+            :param destination: The VOSpace location to put the file to or the
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: VOSpace?

File: vos/vos/storage_inventory.py [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+            :type destination: Node
+            :param send_md5: Should copy send back the md5 of the destination
+            file or just the size?
+            :type send_md5: bool
+            :param disposition: Should the filename from content disposition be
+            returned instead of size or MD5?
+            :type disposition: bool
+            :param head: Return just the headers of a file.
+            :type head: bool
+            :raises When a network problem occurs, it raises one of the
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: Is cutout an option or not. Or it will be added later?

File: vos/vos/storage_inventory.py [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+            raise ValueError('Unable to process server to server copying.')
+        elif is_source_remote:
+            logging.debug(
+                'Source is remote.  Downloading from {} to {}'.format(
+                    source, destination))
+            destination_size = self._get(source, destination)
+            if send_md5:
+                return md5_cache.MD5Cache.compute_md5(destination)
+            elif disposition:
+                # TODO: Should this return the name of the destination instead?
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: TODO?

File: vos/vos/storage_inventory.py [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+
+    def _get(self, source, destination):
+        """
+        :param source: The source Artifact to download (URI).
+        :type uri:  unicode
+        :param destination: The destination to write the data to (path).
+        :type destination: unicode
+        :return: destination size in bytes
+        :rtype long
+        """
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: This, unlike vos, transfers bytes automatically even if source and destination are same?

File: vos/vos/storage_inventory.py [hunk 17]
```diff
@@ -0,0 +1,315 @@
...
+
+        upload_url = '{}/{}'.format(
+            self._get_ws_client()._get_url((STANDARD_ID, None)), destination)
+        stream = Stream(self.conn)
+        logger.debug('Uploading {}'.format(source))
+        return stream.upload(upload_url, destination, source, source_md5,
+                             self.get_metadata)
+
+    def _get_protocol(self):
+        if self.secure_get:
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: We're just dealing with an issue right now where HTTPS is faster than HTTP. The conclusion was to have clients send the protocols they support (in this case both https and http) and let the server decide which one to pick.

File: vos/vos/storage_inventory.py [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+            return 'https'
+        else:
+            return 'http'
+
+    def _is_remote(self, uri):
+        _uri = urlparse(uri)
+        return _uri.scheme and _uri.scheme != 'file'
+
+    def _get_ws_client(self, session_headers=None):
+        return net.BaseWsClient(self.resource_id, net.Subject(), VOS_AGENT,
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: It would be more efficient to cache this for subsequent calls.

File: vos/vos/storage_inventory.py [hunk 19]
```diff
@@ -0,0 +1,315 @@
...
+            'content_encoding': headers.get('Content-Encoding'),
+            'content_length': headers.get('Content-Length'),
+            'content_md5': headers.get('Content-MD5'),
+            'content_type': headers.get('Content-Type')
+        }
+
+    def isdir(self, uri):
+        """
+        Check to see if the given uri is or is a link to a containerNode.
+
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: Is this leftover from vos? All the uris are files in storage, aren't they?

File: vos/vos/storage_inventory.py [hunk 20]
```diff
@@ -0,0 +1,315 @@
...
+            if not _uri.scheme or _uri.scheme == 'file':
+                return os.path.isdir(path) or path.endswith('/')
+            else:
+                return path.endswith('/')
+        except exceptions.NotFoundException:
+            return False
+
+    def isfile(self, uri):
+        """
+        Check if the given uri is or is a link to a DataNode
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: link to a DataNode?

File: vos/vos/storage_inventory.py [hunk 21]
```diff
@@ -0,0 +1,315 @@
...
+
+        :param uri: the VOSpace Node URI to test.
+        :rtype: bool
+        """
+        try:
+            _uri = urlparse(uri)
+            path = _uri.path
+
+            if not _uri.scheme or _uri.scheme == 'file':
+                return os.path.isfile(path) or not path.endswith('/')
```
Comment (andamian) [inline vos/vos/storage_inventory.py]: You might also need to check if it's not a directory. os.path.isdir('/tmp') returns True

File: vos/vos/tests/test_storage_inventory.py [hunk 22]
```diff
@@ -0,0 +1,279 @@
...
+
+class Object(object):
+    pass
+
+
+class TestClient(unittest.TestCase):
+    """Test the vos Client class.
+    """
+
+    TEST_SERVICE_RESOURCE_ID = 'ivo://cadc.nrc.ca/vault'
```
Comment (andamian) [inline vos/vos/tests/test_storage_inventory.py]: vault?

File: vos/vos/tests/test_storage_inventory.py [hunk 23]
```diff
@@ -0,0 +1,279 @@
...
+        # repeat - local file and vospace file are now the same -> only
+        # get_node is called to get the md5 of remote file
+        computed_md5_mock.reset_mock()
+        props.reset_mock()
+        props.get.return_value = md5sum
+        test_client.copy(storageLocation, osLocation)
+        computed_md5_mock.assert_called_once_with(osLocation)
+
+        # change the content of local files to trigger a new copy
+        # computed_md5_mock.reset_mock()
```
Comment (andamian) [inline vos/vos/tests/test_storage_inventory.py]: Tests of the results is the same with the case above when copy was not triggered.

In general these tests seem to just test that the method returns successfully or not and not the actual behaviour.

File: vos/vos/vos.py [hunk 24]
```diff
@@ -2737,3 +2777,291 @@ def get_job_status(self, url):
         """
         return VOFile(url, self.conn, method="GET",
                       follow_redirect=False).read()
+
+
+class Transfer(object):
```
Comment (andamian) [inline vos/vos/vos.py]: I would probably put this in its own module where all other common stuff (storage and vos) should go. utils.py or just 'transfer.py`?

File: vos/vos/vos.py [hunk 25]
```diff
@@ -2737,3 +2777,291 @@ def get_job_status(self, url):
...
+        # follow that link.
+        target = re.search('Unsupported link target:(?P<target> .*)$',
+                           error_message)
+        if target is not None:
+            return target.group('target').strip()
+        raise OSError(error_codes.get(error_message, errno.EFAULT),
+                      '{0}: {1}'.format(uri, error_message))
+
+
+class Stream(object):
```
Comment (andamian) [inline vos/vos/vos.py]: Old item which I don't know if or how was resolved:
Just wondering if this class should encompass more functionality from vos client and ensure the integrity of the transfer (and possible optimizations)

File: vos/new-commands-README.md [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+-i, --interrogate | ask before overwriting files |
+--quick | assuming CANFAR VOSpace, only compatible with CANFAR VOSpace.|
+-L, --follow-links | follow symbolic links. Default is to not follow links. |
+--ignore | ignore errors and continue with recursive copy |
+--head | copy only the headers of a file from vospace. Might return an error if the server does not support the operation ona given file type. |
+
+Examples:
+To reduce clutter, optional arguments are not included in the examples.
+
+Put a local file foo.fits into cadc:TEST. Please note that the '.' at the end will be required
```
[the remark this answers is not in the document]
Comment (yeunga) [reply vos/new-commands-README.md]: For a previous example. Removed the last sentence, i.e. "Please..."

File: vos/new-commands-README.md [hunk 27]
```diff
@@ -0,0 +1,99 @@
+## vostools 3.1.0
```
[the remark this answers is not in the document]
Comment (yeunga) [reply vos/new-commands-README.md]: This file was created to ensure that we have a common understanding of the new command syntax and semantics prior to starting work on the story. 

Please review vcp and vrm help texts to see if the usage info is sufficient.

File: vos/vos/__init__.py [hunk 28]
```diff
@@ -15,3 +15,4 @@
 
 """
 from .vos import Client, Connection, Node, VOFile  # noqa
```
[the remark this answers is not in the document]
Comment (yeunga) [reply vos/vos/__init__.py]: Looks like # noqa is needed. When #noqa is removed, we get the following:
---
(vostools-s2614a-2.7) yeunga@mach276:~/work/git/s2614a/opencadc/vostools/vos/vos$ flake8 __init__.py
__init__.py:18:1: F401 '.storage_inventory.Client as StorageInventoryClient' imported but unused
---

File: vos/vos/commands/tests/test_vrm.py [hunk 29]
```diff
@@ -0,0 +1,250 @@
...
+        parser = CommonParser()
+        parser.add_argument(
+            "--resource-id", default=None,
+            help="resource ID of the Storage Inventory service to be used")
+        parser.add_argument(
+            'source',
+            help='file, dataNode or linkNode to delete from VOSpace',
+            nargs='+')
+        args = parser.parse_args()
+        delete_nodes(args)
```
[the remark this answers is not in the document]
Comment (yeunga) [reply vos/vos/commands/tests/test_vrm.py]: This test case tests the following block (happy path for islink):

----
            if client.get_node(node).islink():
                logging.info('deleting link {}'.format(node))
                client.delete(node)
 ----

Updated the test to assert that client.delete() is called once.

Similarly for isfile.

File: vos/vos/commands/tests/test_vrm.py [hunk 30]
```diff
@@ -0,0 +1,250 @@
...
+        """Test the delete_nodes function in vrm"""
+
+        # happy path, islink
+        mock_node = MagicMock()
+        mock_node.islink.return_value = True
+        client_mock.return_value.get_node.return_value = mock_node
+        client_mock.return_value.delete_nodes.return_value = None
+        sys.argv = [
+            'vrm', '--certfile',
+            '/usr/cadc/local/../dev/admin/test-certificates/\
```
[the remark this answers is not in the document]
Comment (yeunga) [reply vos/vos/commands/tests/test_vrm.py]: Removed the --certfile option, not need for this unit test.

File: vos/vos/storage_inventory.py (hunk 21 above)
Comment (yeunga) [reply vos/vos/storage_inventory.py]: Added check.

File: vos/vos/tests/test_storage_inventory.py (hunk 22 above)
Comment (yeunga) [reply vos/vos/tests/test_storage_inventory.py]: Changed to 'ivo://cadc.nrc.ca/minoc'

File: vos/vos/tests/test_storage_inventory.py (hunk 23 above)
Comment (yeunga) [reply vos/vos/tests/test_storage_inventory.py]: Added file size check.

File: vos/vos/vos.py (hunk 25 above)
Comment (yeunga) [reply vos/vos/vos.py]: Not addressed at this time.

File: vos/vos/vos.py (hunk 24 above)
Comment (yeunga) [reply vos/vos/vos.py]: See technical debt TODO note in Transfer.get_transfer_error() w.r.t. VOFile.

**Pull request merged**